cloudinit/sources/DataSourceIBMCloud.py

```diff
@@ -325,81 +334,82 @@ def read_md():
     return ret
 
 
-def metadata_from_dir(source_dir):
+def metadata_from_dir(source_dir: str) -> Dict[str, Any]:
     """Walk source_dir extracting standardized metadata.
 
     Certain metadata keys are renamed to present a standardized set of metadata
     keys.
 
     This function has a lot in common with ConfigDriveReader.read_v2 but
-    there are a number of inconsistencies, such key renames and as only
-    presenting a 'latest' version which make it an unlikely candidate to share
+    there are a number of inconsistencies, such as key renames and only
+    presenting a 'latest' version, which make it an unlikely candidate to share
     code.
 
     @return: Dict containing translated metadata, userdata, vendordata,
         networkdata as present.
     """
 
-    def opath(fname):
+    def opath(fname: str) -> str:
         return os.path.join("openstack", "latest", fname)
 
-    def load_json_bytes(blob):
+    def load_json_bytes(blob: bytes) -> Dict[str, Any]:
         return json.loads(blob.decode("utf-8"))
 
+    def load_file(
+        path: str, translator: Callable[[bytes], Any]
+    ) -> Any:
+        try:
+            raw = util.load_binary_file(path)
+            return translator(raw)
+        except IOError as e:
+            LOG.debug("Failed reading path '%s': %s", path, e)
+            return None
+        except Exception as e:
+            raise sources.BrokenMetadata(f"Failed decoding {path}: {e}")
+
     files = [
         # tuples of (results_name, path, translator)
         ("metadata_raw", opath("meta_data.json"), load_json_bytes),
-        ("userdata", opath("user_data"), None),
+        ("userdata", opath("user_data"), lambda x: x),
         ("vendordata", opath("vendor_data.json"), load_json_bytes),
         ("networkdata", opath("network_data.json"), load_json_bytes),
     ]
 
-    results = {}
-    for (name, path, transl) in files:
-        fpath = os.path.join(source_dir, path)
-        raw = None
-        try:
-            raw = util.load_binary_file(fpath)
-        except IOError as e:
-            LOG.debug("Failed reading path '%s': %s", fpath, e)
-
-        if raw is None or transl is None:
-            data = raw
-        else:
-            try:
-                data = transl(raw)
-            except Exception as e:
-                raise sources.BrokenMetadata(
-                    "Failed decoding %s: %s" % (path, e)
-                )
+    results: Dict[str, Any] = {}
 
-        results[name] = data
+    for name, path, transl in files:
+        fpath = os.path.join(source_dir, path)
+        results[name] = load_file(fpath, transl)
 
-    if results.get("metadata_raw") is None:
+    if results["metadata_raw"] is None:
         raise sources.BrokenMetadata(
-            "%s missing required file 'meta_data.json'" % source_dir
+            "%s missing required file 'meta_data.json'",
```

Comment:
The comma syntax only works for logging because the logging module has some smarts to not interpolate the strings if it's not gonna output the log anywhere.

`%` is correct here, but since we're modifying it anyway, we could update it to be an f-string instead.